### PR TITLE
Add jetty support

### PR DESCRIPTION
**Pull request by tokuhirom**

### CODE REVIEW

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  private static final List<String> EMPTY_LIST = new ArrayList<String>();
+
+  public JettyStatisticsCollector(StatisticsHandler statisticsHandler) {
+    this.statisticsHandler = statisticsHandler;
+  }
+
+  @Override
+  public List<MetricFamilySamples> collect() {
+    return Arrays.asList(
+            buildGauge("jetty_requests", "number of requests", statisticsHandler.getRequests()),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: What sort of requests?

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            buildGauge("jetty_requests_active", "number of requests currently active", statisticsHandler.getRequestsActive()),
+            buildGauge("jetty_requests_active_max", "maximum number of active requests", statisticsHandler.getRequestsActiveMax()),
+            buildGauge("jetty_request_time_max", "maximum time spend handling requests (in ms)", statisticsHandler.getRequestTimeMax()),
+            buildGauge("jetty_request_time_total", "total time spend in all request handling (in ms)", statisticsHandler.getRequestTimeTotal()),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: Units should be in the metric name, and we prefer seconds

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            buildGauge("jetty_request_time_mean", "mean time spent handling requests (in ms)", statisticsHandler.getRequestTimeMean()),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: These two aren't useful, drop them

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  public JettyStatisticsCollector(StatisticsHandler statisticsHandler) {
+    this.statisticsHandler = statisticsHandler;
+  }
+
+  @Override
+  public List<MetricFamilySamples> collect() {
+    return Arrays.asList(
+            buildGauge("jetty_requests", "number of requests", statisticsHandler.getRequests()),
+            buildGauge("jetty_requests_active", "number of requests currently active", statisticsHandler.getRequestsActive()),
+            buildGauge("jetty_requests_active_max", "maximum number of active requests", statisticsHandler.getRequestsActiveMax()),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: Is this a limit or a watermark?

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            buildGauge("jetty_dispatched_time_total", "total time spent in dispatch handling (in ms)", statisticsHandler.getDispatchedTimeTotal()),
+            buildGauge("jetty_dispatched_time_mean", "mean time spent in dispatch handling (in ms)", statisticsHandler.getDispatchedTimeMean()),
+            buildGauge("jetty_dispatched_time_std_dev", "standard deviation for dispatch handling (in ms)", statisticsHandler.getDispatchedTimeStdDev()),
+            buildGauge("jetty_async_requests", "total number of async requests", statisticsHandler.getAsyncRequests()),
+            buildGauge("jetty_async_requests_waiting", "currently waiting async requests", statisticsHandler.getAsyncRequestsWaiting()),
+            buildGauge("jetty_async_requests_waiting_max", "maximum number of waiting async requests", statisticsHandler.getAsyncRequestsWaitingMax()),
+            buildGauge("jetty_async_dispatches", "number of requested that have been asynchronously dispatched", statisticsHandler.getAsyncDispatches()),
+            buildGauge("jetty_expires", "number of async requests requests that have expired", statisticsHandler.getExpires()),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: This sounds like a counter, so it should end in _total.

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                    buildStatusSample(name, "4xx", statisticsHandler.getResponses4xx()),
+                    buildStatusSample(name, "5xx", statisticsHandler.getResponses5xx())
+            )
+    );
+  }
+
+  private static MetricFamilySamples.Sample buildStatusSample(String name, String status, double value) {
+    return new MetricFamilySamples.Sample(
+            name,
+            Collections.singletonList("status"),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: We usually call this `code`

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  private static MetricFamilySamples buildGauge(String name, String help, double value) {
+    return new MetricFamilySamples(
+            name,
+            Type.GAUGE,
+            help,
+            Collections.singletonList(new MetricFamilySamples.Sample(name, EMPTY_LIST, EMPTY_LIST, value)));
+  }
+
+  private MetricFamilySamples buildStatusGauge() {
+    String name = "jetty_responses";
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: This is only used once, inline.

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 1 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: It's # of  http requests handled by jetty.

Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: This should be a Counter then with a _total suffix, or combine with request_time to produce a summary.

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 4 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: It's a watermark. This is the maximum number of active requests when there were many connections.

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 6 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: fixed.

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 7 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: It's used at https://github.com/prometheus/client_java/pull/120/files#diff-f47a35564451e6d56e9b2ec018b94324R65

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 4 above)
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: The help string could make that clearer: "Maximum number of requests that have been active."

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 3 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: which are these? jetty_request_time_mean and jetty_request_time_total?

Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: mean and stddev

Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: i see.

Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: removed.

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 2 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: Added _seconds suffix and divide value by 1000.0.

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 1 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: Added _total suffix.

**Comment by tokuhirom** [comment]
I rewrote all points you pointed :)

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import io.prometheus.client.Collector;
+import org.eclipse.jetty.server.handler.StatisticsHandler;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Collect metrics from jettty's org.eclipse.jetty.server.handler.StatisticsHandler.
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: Can you put example usage in the docstring?

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+  @Override
+  public List<MetricFamilySamples> collect() {
+    return Arrays.asList(
+            buildGauge("jetty_requests_total", "number of requests", statisticsHandler.getRequests()),
+            buildGauge("jetty_requests_active", "number of requests currently active", statisticsHandler.getRequestsActive()),
+            buildGauge("jetty_requests_active_max", "Maximum number of requests that have been active at once", statisticsHandler.getRequestsActiveMax()),
+            buildGauge("jetty_request_time_max_seconds", "maximum time spend handling requests", statisticsHandler.getRequestTimeMax()/1000.0),
+            buildGauge("jetty_request_time_total_seconds", "total time spend in all request handling", statisticsHandler.getRequestTimeTotal()/1000.0),
+            buildGauge("jetty_dispatched", "number of dispatches", statisticsHandler.getDispatched()),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: This sounds like a counter, so it should end in _total

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  @Override
+  public List<MetricFamilySamples> collect() {
+    return Arrays.asList(
+            buildGauge("jetty_requests_total", "number of requests", statisticsHandler.getRequests()),
+            buildGauge("jetty_requests_active", "number of requests currently active", statisticsHandler.getRequestsActive()),
+            buildGauge("jetty_requests_active_max", "Maximum number of requests that have been active at once", statisticsHandler.getRequestsActiveMax()),
+            buildGauge("jetty_request_time_max_seconds", "maximum time spend handling requests", statisticsHandler.getRequestTimeMax()/1000.0),
+            buildGauge("jetty_request_time_total_seconds", "total time spend in all request handling", statisticsHandler.getRequestTimeTotal()/1000.0),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: _seconds_total as it's a Counter

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            buildGauge("jetty_dispatched", "number of dispatches", statisticsHandler.getDispatched()),
+            buildGauge("jetty_dispatched_active", "number of dispatches currently active", statisticsHandler.getDispatchedActive()),
+            buildGauge("jetty_dispatched_active_max", "maximum number of active dispatches being handled", statisticsHandler.getDispatchedActiveMax()),
+            buildGauge("jetty_dispatched_time_max", "maximum time spend in dispatch handling", statisticsHandler.getDispatchedTimeMax()),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: spent

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            buildGauge("jetty_dispatched_time_total_seconds", "total time spent in dispatch handling", statisticsHandler.getDispatchedTimeTotal()/1000.0),
+            buildGauge("jetty_async_requests", "total number of async requests", statisticsHandler.getAsyncRequests()),
+            buildGauge("jetty_async_requests_waiting", "currently waiting async requests", statisticsHandler.getAsyncRequestsWaiting()),
+            buildGauge("jetty_async_requests_waiting_max", "maximum number of waiting async requests", statisticsHandler.getAsyncRequestsWaitingMax()),
+            buildGauge("jetty_async_dispatches", "number of requested that have been asynchronously dispatched", statisticsHandler.getAsyncDispatches()),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: `_total`

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 8 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: Okay, added.

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 9 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: Added _total suffix

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java (hunk 11 above)
Comment (tokuhirom) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: fixed.

File: simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+  private static final List<String> EMPTY_LIST = new ArrayList<String>();
+
+  public JettyStatisticsCollector(StatisticsHandler statisticsHandler) {
+    this.statisticsHandler = statisticsHandler;
+  }
+
+  @Override
+  public List<MetricFamilySamples> collect() {
+    return Arrays.asList(
+            buildGauge("jetty_requests_total", "number of requests", statisticsHandler.getRequests()),
```
Comment (brian-brazil) [inline simpleclient_jetty/src/main/java/io/prometheus/client/jetty/JettyStatisticsCollector.java]: Could you start all the help strings with a capital letter?

**Comment by tokuhirom** [comment]
Fixed labels and helps.

**Comment by brian-brazil** [comment]
Did you forget to git push?

Presuming that's done this is ready to merge, so could you squash your commits please?

**Comment by tokuhirom** [comment]
Oops. I capitalised first letter of help text.
And pushed squashed commits.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**